### ISSUE TITLE
Helm upgrade deletes namespace and related resources

### ISSUE THREAD

**Issue by DaruZero** [helm/helm#12287]
I am experiencing a weird behavior when performing an helm upgrade. Here's the situation:
I have a simple chart, that is deployed in the default namespace with the usual `helm install mgmt-companion url-to-chart`, and all the resources deployed by the chart end up in the `mgmt-companion` namespace, that I previously created with `kubectl create ns mgmt-companion`. So basically the release is installed in the `default` namespace, and the resources in the `mgmt-companion` namespace.
Now, I am doing some local changes to the chart, and I wanted to test them out before publishing a new version. So I ran the command `helm upgrade mgmt-companion ./path/to/local/chart`, and it exited with the normal Happy Helming success message. The problem is that the `mgmt-companion` namespace, and all the resources within it, started to get terminated. Nothing like this ever happened to me, and I cannot figure out why an upgrade deletes all the resources.
Here's the output of the upgrade command with the `--debug` option:
```
# helm upgrade mgmt-companion .\deployment\mgmt-companion\ --debug
upgrade.go:144: [debug] preparing upgrade for mgmt-companion
upgrade.go:526: [debug] copying values from mgmt-companion (v1) to new release.
upgrade.go:152: [debug] performing update for mgmt-companion
upgrade.go:324: [debug] creating upgraded release for mgmt-companion
client.go:396: [debug] checking 9 resources for changes
client.go:675: [debug] Looks like there are no changes for ServiceAccount "mgmt-companion-helm"
client.go:675: [debug] Looks like there are no changes for Secret "certs-secrets"
client.go:684: [debug] Patch ConfigMap "configmap-datasource" in namespace mgmt-companion
client.go:675: [debug] Looks like there are no changes for ConfigMap "custom-hosts"
client.go:675: [debug] Looks like there are no changes for ClusterRole "mgmt-companion-helm"
client.go:675: [debug] Looks like there are no changes for ClusterRoleBinding "mgmt-companion-helm-clusterrolebinding"
client.go:675: [debug] Looks like there are no changes for RoleBinding "managementconsolecompanion-rolebinding"
client.go:684: [debug] Patch Deployment "managementconsolecompanion" in namespace mgmt-companion
client.go:684: [debug] Patch Deployment "mgmtconsoledatasourcedeployer" in namespace mgmt-companion
client.go:445: [debug] Deleting Namespace "mgmt-companion" in namespace ...
upgrade.go:159: [debug] updating status for upgraded release for mgmt-companion
Release "mgmt-companion" has been upgraded. Happy Helming!
NAME: mgmt-companion
LAST DEPLOYED: Wed Aug  9 14:26:20 2023
NAMESPACE: default
STATUS: deployed
REVISION: 2
TEST SUITE: None
USER-SUPPLIED VALUES:
...
```

The only things changing in the chart are some of the values and the template of a configmap.

Output of `helm version`:
```
version.BuildInfo{Version:"v3.11.2", GitCommit:"912ebc1cd10d38d340f048efaf0abda047c3468e", GitTreeState:"clean", GoVersion:"go1.18.10"}
```
Output of `kubectl version`:
```
WARNING: This version information is deprecated and will be replaced with the output from kubectl version --short.  Use --output=yaml|json to get the full version.
Client Version: version.Info{Major:"1", Minor:"26", GitVersion:"v1.26.1", GitCommit:"8f94681cd294aa8cfd3407b8191f6c70214973a4", GitTreeState:"clean", BuildDate:"2023-01-18T15:58:16Z", GoVersion:"go1.19.5", Compiler:"gc", Platform:"windows/amd64"}
Kustomize Version: v4.5.7
Server Version: version.Info{Major:"1", Minor:"26", GitVersion:"v1.26.4+k3s1", GitCommit:"8d0255af07e95b841952563253d27b0d10bd72f0", GitTreeState:"clean", BuildDate:"2023-04-20T00:33:18Z", GoVersion:"go1.19.8", Compiler:"gc", Platform:"linux/amd64"}
```
Cloud Provider/Platform (AKS, GKE, Minikube etc.): 
k3d

**Comment by Ithrael** [comment]
Can I take a look at your charts code?

**Comment by gjenkins8** [comment]
The reason for the deletion looks to be:
```
client.go:445: [debug] Deleting Namespace "mgmt-companion" in namespace ...
```

The best reason I can think of, is that the chart previously included/created the "mgmt-companion" namespace. Then this namespace was removed from the chart, and now Helm thinks it needs to do the corresponding deletion on the cluster.

Overall, Helm is just adding/removing objects as described by your chart-versions (ie. it doesn't have some sort of built-in "delete the namespace after upgrade" logic)

**Comment by DaruZero** [comment]
Yeah seems like the problem is related to that.
Basically the Chart version installed in my test cluster was `1.0.126`, but the Chart version in my local branch was a little behind to `1.0.124`. So doing the upgrade reverted to a previous version. The problem is that in version `1.0.124` we had a template definition for the namespace, that has been removed in `1.0.126`.

Solving my problem was just a matter of rebasing my branch, but still I don't know why if in the initially installed version`1.0.126` there wasn't a template for the namespace (as it was created with `kubectl`), upgrading to an older version of the chart where the template was present still deleted the namespace resource.

As I managed to solve my issue, we can close this thread, unless you would like to investigate more. Unfortunately I cannot share much of the chart
